File: Skills/markitdown/scripts/convert.ts
#!/usr/bin/env bun
/**
 * MarkItDown Converter - Single file conversion
 * Convert any file to Markdown format
 */

import { parseArgs } from "util";
import { spawn } from "child_process";
import { existsSync } from "fs";
import { basename, extname, join, dirname } from "path";

const { values, positionals } = parseArgs({
  args: Bun.argv.slice(2),
  options: {
    input: { type: "string", short: "i" },
    output: { type: "string", short: "o" },
    extension: { type: "string", short: "x" },
    "mime-type": { type: "string", short: "m" },
    charset: { type: "string", short: "c" },
    plugins: { type: "boolean", short: "p" },
    prompt: { type: "string" },
    help: { type: "boolean", short: "h" },
  },
  allowPositionals: true,
});

if (values.help || (!values.input && positionals.length === 0)) {
  console.log(`
MarkItDown Converter - Convert any file to Markdown

Usage:
  bun convert.ts --input <file> [options]
  bun convert.ts <file> [options]

Options:
  -i, --input <path>      Input file path (required if not positional)
  -o, --output <path>     Output file path (default: stdout)
  -x, --extension <ext>   Force file extension hint (e.g., "pdf")
  -m, --mime-type <type>  Force MIME type hint
  -c, --charset <charset> Force charset hint (e.g., "UTF-8")
  -p, --plugins           Enable 3rd-party plugins
  --prompt <text>         Custom prompt for image/audio descriptions
  -h, --help              Show this help

Examples:
  bun convert.ts --input report.pdf --output report.md
  bun convert.ts presentation.pptx -o notes.md
  bun convert.ts --input document.docx --plugins
`);
  process.exit(0);
}

const inputPath = values.input || positionals[0];

if (!inputPath) {
  console.error("Error: Input file is required");
  process.exit(1);
}

if (!existsSync(inputPath)) {
  console.error(`Error: File not found: ${inputPath}`);
  process.exit(1);
}

// Build markitdown command arguments
const args: string[] = [inputPath];

if (values.output) {
  args.push("-o", values.output);
}

if (values.extension) {
  args.push("-x", values.extension);
}

if (values["mime-type"]) {
  args.push("-m", values["mime-type"]);
}

if (values.charset) {
  args.push("-c", values.charset);
}

if (values.plugins) {
  args.push("-p");
}

// Execute markitdown
const proc = spawn("markitdown", args, {
  stdio: ["inherit", "pipe", "pipe"],
});

let stdout = "";
let stderr = "";

proc.stdout.on("data", (data) => {
  stdout += data.toString();
});

proc.stderr.on("data", (data) => {
  stderr += data.toString();
});

proc.on("close", (code) => {
  if (code !== 0) {
    console.error(`Conversion failed (exit ${code}):`);
    if (stderr) console.error(stderr);
    process.exit(code || 1);
  }

  // If output was specified, markitdown already wrote to file
  // If not, we need to output the result
  if (!values.output) {
    console.log(stdout);
  } else {
    console.log(`✓ Converted: ${inputPath} → ${values.output}`);
  }
});

File: Skills/markitdown/scripts/convert.py
#!/usr/bin/env python3
"""
MarkItDown conversion script for Zo Skill.
Converts any file to Markdown for LLM consumption.
"""

import argparse
import sys
import os
from pathlib import Path

# Import after installation check
try:
    from markitdown import MarkItDown
except ImportError:
    print("Error: markitdown not installed. Run: pip install 'markitdown[all]'", file=sys.stderr)
    sys.exit(1)


def main():
    parser = argparse.ArgumentParser(
        description="Convert any file to Markdown using MarkItDown"
    )
    parser.add_argument("input", help="Input file path or URL")
    parser.add_argument("output", nargs="?", help="Output markdown file (optional, defaults to stdout)")
    parser.add_argument("--use-plugins", action="store_true", help="Enable 3rd-party plugins")
    parser.add_argument("--docintel", action="store_true", help="Use Azure Document Intelligence")
    parser.add_argument("--endpoint", help="Azure Document Intelligence endpoint")
    parser.add_argument("--llm-api-key", help="OpenAI API key for image descriptions")
    parser.add_argument("--llm-model", default="gpt-4o", help="LLM model for image descriptions")
    parser.add_argument("--list-plugins", action="store_true", help="List available plugins")
    parser.add_argument("--verbose", "-v", action="store_true", help="Verbose output")
    
    args = parser.parse_args()
    
    # Initialize MarkItDown
    md_kwargs = {"enable_plugins": args.use_plugins}
    
    # Add Azure Document Intelligence if requested
    if args.docintel:
        if not args.endpoint:
            print("Error: --endpoint required when using --docintel", file=sys.stderr)
            sys.exit(1)
        md_kwargs["docintel_endpoint"] = args.endpoint
    
    # Add LLM client if API key provided
    if args.llm_api_key and args.llm_model:
        try:
            from openai import OpenAI
            md_kwargs["llm_client"] = OpenAI(api_key=args.llm_api_key)
            md_kwargs["llm_model"] = args.llm_model
        except ImportError:
            print("Warning: openai package not installed, LLM features disabled", file=sys.stderr)
    
    md = MarkItDown(**md_kwargs)
    
    # List plugins if requested
    if args.list_plugins:
        plugins = md._plugins if hasattr(md, '_plugins') else []
        if plugins:
            print("Available plugins:")
            for plugin in plugins:
                print(f"  - {plugin}")
        else:
            print("No plugins available/loaded")
        sys.exit(0)
    
    # Check input file exists (if not a URL)
    if not args.input.startswith("http") and not os.path.exists(args.input):
        print(f"Error: Input file not found: {args.input}", file=sys.stderr)
        sys.exit(1)
    
    # Convert file
    if args.verbose:
        print(f"Converting: {args.input}...", file=sys.stderr)
    
    try:
        result = md.convert(args.input)
    except Exception as e:
        print(f"Error during conversion: {e}", file=sys.stderr)
        sys.exit(1)
    
    # Output result
    output = result.text_content
    
    if args.output:
        with open(args.output, 'w', encoding='utf-8') as f:
            f.write(output)
        if args.verbose:
            print(f"Written to: {args.output}", file=sys.stderr)
    else:
        print(output)


if __name__ == "__main__":
    main()
